codex/sales/reservations.nim

```diff
@@ -0,0 +1,314 @@
+
+func used*(availability: Availability): bool =
+  availability.slotId.isSome
+
+func key(id: AvailabilityId): ?!Key =
+  (ReservationsKey / id.toArray.toHex)
+
+func key*(availability: Availability): ?!Key =
+  return availability.id.key
+
+func available*(self: Reservations): uint =
+  return self.repo.quotaMaxBytes - self.repo.totalUsed
+
+func available*(self: Reservations, bytes: uint): bool =
+  return bytes < self.available()
+
+proc exists*(
+  self: Reservations,
+  id: AvailabilityId): Future[?!bool] {.async.} =
+
+  without key =? id.key, err:
+    return failure(err)
+
+  let exists = await self.repo.metaDs.contains(key)
+  return success(exists)
+
+proc get*(
+  self: Reservations,
+  id: AvailabilityId): Future[?!Availability] {.async.} =
+
+  if exists =? (await self.exists(id)) and not exists:
+    let err = newException(AvailabilityNotExistsError,
+      "Availability does not exist")
+    return failure(err)
+
+  without key =? id.key, err:
+    return failure(err)
+
+  without serialized =? await self.repo.metaDs.get(key), err:
+    return failure(err)
+
+  without availability =? Json.decode(serialized, Availability).catch, err:
+    return failure(err)
+
+  return success availability
+
+proc update(
+  self: Reservations,
+  availability: Availability,
+  slotId: ?SlotId): Future[?!void] {.async.} =
+
+  without var updated =? await self.get(availability.id), err:
+    return failure(err)
+
+  updated.slotId = slotId
+
+  without key =? availability.key, err:
+    return failure(err)
+
+  if err =? (await self.repo.metaDs.put(
+    key,
+    @(updated.toJson.toBytes))).errorOption:
+    return failure(err)
+
+  return success()
+
+proc reserve*(
+  self: Reservations,
+  availability: Availability): Future[?!void] {.async.} =
+
+  if exists =? (await self.exists(availability.id)) and exists:
+    let err = newException(AvailabilityAlreadyExistsError,
+      "Availability already exists")
+    return failure(err)
+
+  without key =? availability.key, err:
+    return failure(err)
+
+  if err =? (await self.repo.metaDs.put(
+    key,
+    @(availability.toJson.toBytes))).errorOption:
+    return failure(err)
+
+  # TODO: reconcile data sizes -- availability uses UInt256 and RepoStore
+  # uses uint, thus the need to truncate
+  if reserveInnerErr =? (await self.repo.reserve(
+    availability.size.truncate(uint))).errorOption:
+
+    let reserveErr = reserveInnerErr.toErr(AvailabilityReserveFailedError,
+      "Availability reservation failed")
+
+    # rollback persisted availability
+    if rollbackInnerErr =? (await self.repo.metaDs.delete(key)).errorOption:
+      let rollbackErr = rollbackInnerErr.toErr(AvailabilityDeleteFailedError,
+        "Failed to delete persisted availability during rollback")
+      rollbackInnerErr.parent = reserveErr
+      return failure(rollbackErr)
+
+    return failure(reserveErr)
+
+  return success()
+
+# TODO: call site not yet determined. Perhaps reuse of Availabilty should be set
```

Comment:
The "list of availability" model is starting to show its cracks here. We should probably move to a different model soon, whereby we allow the user to simply indicate once how much storage they want to make available in total, and for which duration and price they'd like to engage into contracts. It would be up to the sales agent to sell off pieces of the total available storage as it sees fit.

Reply:
This is indeed how I envisioned it and how it got implemented in the RepoStore.

Reply:
> The "list of availability" model is starting to show its cracks here. We should probably move to a different model soon, whereby we allow the user to simply indicate once how much storage they want to make available in total, and for which duration and price they'd like to engage into contracts. It would be up to the sales agent to sell off pieces of the total available storage as it sees fit.

Should we handle this as a future PR so that we can move forward with user testing?

> This is indeed how I envisioned it and how it got implemented in the RepoStore.

Agreed. We discussed moving to a different model, but wanted to retrofit this to the old availability model so we could enable user testing, with plans to update the model and api.

Reply:
Agree, what was the reasoning for having this parameter mutable and exposed over REST? IMHO I would expect having something like @markspanbroek is describing and having it in CLI/config parameter...

Reply:
> having something like @markspanbroek is describing and having it in CLI/config parameter...

I agree with this, it always felt odd being a REST endpoint. Perhaps it made more sense to have a REST endpoint in the old model as it contained a list of availabilities. The other benefit of having a REST endpoint would be modifying the availability without restarting the process. 

Moving forward with the new model as described it would make a lot of sense to include available bytes, duration, and min cost as CLI parameters, and then later if we want the ability to modify these values dynamically, without process restart, we can add that it in. Wdyt?

Reply:
> Moving forward with the new model as described it would make a lot of sense to include available bytes, duration, and min cost as CLI parameters, and then later if we want the ability to modify these values dynamically, without process restart, we can add that it in. Wdyt?

Yes, I agree. The REST endpoint made sense to me for the old model, but it no longer makes sense in the new model. When we implement the new model (in a different PR, I'd say), we should move to CLI/config parameters.

Reply:
Let's keep the availabilities via REST for now as discussed in https://github.com/status-im/codex-research/pull/161.